protocol/verification.go

```diff
@@ -0,0 +1,104 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	"github.com/coniks-sys/coniks-go/merkletree"
+)
+
+func (res *ErrorResponse) Verify(string, []byte, sign.PublicKey) ErrorCode {
+	return res.Error
+}
+
+func (df *DirectoryProof) Verify(uname string, savedSTR []byte, signKey sign.PublicKey) ErrorCode {
+	// step 1. verify TB's signature
+	if df.TB != nil {
+		if !signKey.Verify(df.TB.Serialize(df.STR.Signature), df.TB.Signature) {
+			return ErrorBadSignature
+		}
+	}
+
+	// step 2. verify STR's signature
```

Comment:
I think it makes more sense to verify the STR signature before verifying the TB's signature, since the TB's signature depends on the STR's signature.